packages/actor-rdf-resolve-hypermedia-links-queue-wrapper-info-occupancy/lib/LinkQueueLogger.ts

```diff
@@ -0,0 +1,186 @@
+import type { ILinkQueue, ILink } from '@comunica/bus-rdf-resolve-hypermedia-links-queue';
+import { LinkQueueWrapper } from '@comunica/bus-rdf-resolve-hypermedia-links-queue';
+import type { Logger } from '@comunica/types';
+
+/**
+ * A link queue that log information about event happening in the queue
+ */
+export class LinkQueueLogger extends LinkQueueWrapper {
+  public readonly query: string;
+  private readonly logger: Logger;
+  private readonly linkProductionRatio: ILinkProductionActorRatio = {
+    pushEvent: {},
+    popEvent: {},
+  };
+
+  public static readonly LINK_QUEUE_EVENT_NAME = '<Link queue occupancy>';
+  public static readonly LINK_QUEUE_DIDNT_START_EMPTY_MESSAGE = 'the link queue didn\'t start empty';
+
+  /**
+   *
+   * @param {ILinkQueue} linkQueue - The link queue
+   * @param {Algebra.Operation} query - The current query
+   * @param {Logger} logger - The logger where the events are output
+   */
+  public constructor(linkQueue: ILinkQueue, query: string, logger: Logger) {
+    super(linkQueue);
+    this.logger = logger;
+
+    if (!linkQueue.isEmpty()) {
+      this.logger.warn(
+        LinkQueueLogger.LINK_QUEUE_EVENT_NAME,
+        {
+          message: LinkQueueLogger.LINK_QUEUE_DIDNT_START_EMPTY_MESSAGE,
+        },
+      );
+    }
+    this.query = query;
+
+    Object.freeze(this.query);
+  }
+
+  /**
+   * Helper function to get the reachability criteria of a link
+   * @param {ILink} link - Current link
+   * @returns {IProduceByActor | null } The reachability criteria with extra information about it if available
+   */
+  private static getActorProductorInformation(link: ILink): IProduceByActor | null {
+    const metadata = link.metadata;
+    if (metadata !== undefined && metadata[PRODUCED_BY_ACTOR] !== undefined) {
+      const { name, ...rest } = metadata[PRODUCED_BY_ACTOR];
+
+      if (name === undefined) {
+        return null;
+      }
+      return {
+        name,
+        metadata: Object.keys(rest).length === 0 ? undefined : rest,
+      };
+    }
+    return null;
+  }
+
+  /**
+   * Update the link production ratio
+   * @param {IURLStatistic} link - current link
+   * @param {keyof ILinkProductionActorRatio} event - link queue event
+   */
+  private updateLinkProductionRatio(link: IURLStatistic, event: keyof ILinkProductionActorRatio): void {
+    if (link.producedByActor !== null &&
+      this.linkProductionRatio[event][link.producedByActor?.name] !== undefined) {
+      this.linkProductionRatio[event][link.producedByActor?.name] += 1;
+    } else if (link.producedByActor !== null &&
+      this.linkProductionRatio[event][link.producedByActor?.name] === undefined) {
+      this.linkProductionRatio[event][link.producedByActor?.name] = 1;
```

Comment:
If the `!this.linkProductionRatio[event][link.producedByActor?.name]` check is done first, it can be initialized to zero. Then, this if statement won't be needed anymore.

(similar for the code below)